File: Src/java/cql-to-elm/src/jvmTest/java/org/cqframework/cql/cql2elm/ElmSupportTest.java
package org.cqframework.cql.cql2elm;

import static kotlinx.io.CoreKt.buffered;
import static kotlinx.io.JvmCoreKt.asSource;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import org.cqframework.cql.cql2elm.LibraryBuilder.SignatureLevel;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

class ElmSupportTest {

    static ModelManager modelManager;
    static LibraryManager libraryManager;

    @BeforeAll
    static void setup() {
        modelManager = new ModelManager();
        libraryManager = new LibraryManager(modelManager);
        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());
    }

    @AfterAll
    static void tearDown() {
        libraryManager.getLibrarySourceLoader().clearProviders();
    }

    @Test
    void includedLibraryWithJsonElm() {
        CqlCompilerOptions options =
                new CqlCompilerOptions(CqlCompilerException.ErrorSeverity.Info, SignatureLevel.All);
        libraryManager = new LibraryManager(modelManager, options);

        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());
        try {

            var translator = CqlTranslator.fromSource(
                    buffered(asSource(LibraryTests.class.getResourceAsStream("LibraryTests/ReferencingLibraryJsonElm.cql"))),
                    libraryManager);

            assertTrue(translator.getErrors().size() > 0);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void includedLibraryWithXmlElm() {
        CqlCompilerOptions options =
                new CqlCompilerOptions(CqlCompilerException.ErrorSeverity.Info, SignatureLevel.All);
        libraryManager = new LibraryManager(modelManager, options);
        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());

        try {
            var translator = CqlTranslator.fromSource(
                    buffered(asSource(LibraryTests.class.getResourceAsStream("LibraryTests/ReferencingLibraryXmlElm.cql"))),
                    libraryManager);

            assertTrue(translator.getErrors().size() > 0);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    void includedLibraryWithJsonWithNullTypeSpecifierElm() {
        CqlCompilerOptions options =
                new CqlCompilerOptions(CqlCompilerException.ErrorSeverity.Info, SignatureLevel.All);
        libraryManager = new LibraryManager(modelManager, options);
        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());
        try {
            var translator = CqlTranslator.fromSource(
                    buffered(asSource(LibraryTests.class.getResourceAsStream(
                            "LibraryTests/ReferencingLibraryWithNullTypeSpecifierJsonElm.cql"))),
                    libraryManager);

            assertTrue(translator.getErrors().size() > 0);

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private CqlCompilerOptions createOptions() {
        CqlCompilerOptions result = new CqlCompilerOptions();
        result.setOptions(
                CqlCompilerOptions.Options.EnableDateRangeOptimization,
                CqlCompilerOptions.Options.EnableAnnotations,
                CqlCompilerOptions.Options.EnableLocators,
                CqlCompilerOptions.Options.EnableResultTypes,
                CqlCompilerOptions.Options.DisableListDemotion,
                CqlCompilerOptions.Options.DisableListPromotion,
                CqlCompilerOptions.Options.DisableMethodInvocation);

        return result;
    }
}

File: Src/java/cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/ElmSupportTest.kt
package org.cqframework.cql.cql2elm

import java.io.IOException
import kotlinx.io.asSource
import kotlinx.io.buffered
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

@Suppress("PrintStackTrace")
internal class ElmSupportTest {
    @Test
    fun includedLibraryWithJsonElm() {
        val options =
            CqlCompilerOptions(
                CqlCompilerException.ErrorSeverity.Info,
                LibraryBuilder.SignatureLevel.All,
            )
        libraryManager = LibraryManager(modelManager!!, options)

        libraryManager!!.librarySourceLoader.registerProvider(TestLibrarySourceProvider())
        try {
            val translator =
                CqlTranslator.fromSource(
                    LibraryTests::class
                        .java
                        .getResourceAsStream("LibraryTests/ReferencingLibraryJsonElm.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )

            Assertions.assertTrue(translator.errors.isNotEmpty())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun includedLibraryWithXmlElm() {
        val options =
            CqlCompilerOptions(
                CqlCompilerException.ErrorSeverity.Info,
                LibraryBuilder.SignatureLevel.All,
            )
        libraryManager = LibraryManager(modelManager!!, options)
        libraryManager!!.librarySourceLoader.registerProvider(TestLibrarySourceProvider())

        try {
            val translator =
                CqlTranslator.fromSource(
                    LibraryTests::class
                        .java
                        .getResourceAsStream("LibraryTests/ReferencingLibraryXmlElm.cql")!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )

            Assertions.assertTrue(translator.errors.isNotEmpty())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    @Test
    fun includedLibraryWithJsonWithNullTypeSpecifierElm() {
        val options =
            CqlCompilerOptions(
                CqlCompilerException.ErrorSeverity.Info,
                LibraryBuilder.SignatureLevel.All,
            )
        libraryManager = LibraryManager(modelManager!!, options)
        libraryManager!!.librarySourceLoader.registerProvider(TestLibrarySourceProvider())
        try {
            val translator =
                CqlTranslator.fromSource(
                    LibraryTests::class
                        .java
                        .getResourceAsStream(
                            "LibraryTests/ReferencingLibraryWithNullTypeSpecifierJsonElm.cql"
                        )!!
                        .asSource()
                        .buffered(),
                    libraryManager!!,
                )

            Assertions.assertTrue(translator.errors.isNotEmpty())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        var modelManager: ModelManager? = null
        var libraryManager: LibraryManager? = null

        @JvmStatic
        @BeforeAll
        fun setup() {
            modelManager = ModelManager()
            libraryManager = LibraryManager(modelManager!!)
            libraryManager!!.librarySourceLoader.registerProvider(TestLibrarySourceProvider())
        }

        @JvmStatic
        @AfterAll
        fun tearDown() {
            libraryManager!!.librarySourceLoader.clearProviders()
        }
    }
}
